### DISCUSSION TITLE
Store internal states of speculative decode

### DISCUSSION THREAD

**Discussion by vladislavkruglikov** [vllm-project/vllm#7522] [General]
I want to get internal states of speculative decode such as history of proposals and scores in the response as well as completion

To achieve my goal I generally need to get proposals and proposal scores from model executor

https://github.com/vllm-project/vllm/blob/d3d9cb6e4b8185b4e56e1dda92c6fc31cdc05de1/vllm/engine/async_llm_engine.py#L282-L283

After this pass to function that is responsible for update of internal state of sequence with outputs from model executor

https://github.com/vllm-project/vllm/blob/d3d9cb6e4b8185b4e56e1dda92c6fc31cdc05de1/vllm/engine/async_llm_engine.py#L287-L289

But the problem that I face is that list[SamplerOutput] that is being returned from the model executor and speculative worker under the hood stores data about each token not a whole sequence. I mean I could possibly append proposals and proposal scores for every token in sampler output but this would be just weird and scary so I need to figure out some prettier way to do that

One way is to enrich return type of worker to allow it store some metadata except sampler output for example something like this would work

```python
class WorkerResponse:
    sampler_outputs: list[SamplerOutput]

class SpeculativeWorkerResponse(WorkerResponse):
    proposals: SpeculativeProposals
    proposal_scores: SpeculativeScores
```

So change from this

https://github.com/vllm-project/vllm/blob/d3d9cb6e4b8185b4e56e1dda92c6fc31cdc05de1/vllm/spec_decode/spec_decode_worker.py#L333-L337

To this

```python
@torch.inference_mode()
def execute_model(
    self,
    execute_model_req: Optional[ExecuteModelRequest] = None
) -> SpeculativeWorkerResponse:
```

This data structure can be returned from speculative worker and after all returned by model executor and passed to post process function that would append proposals and proposal scores to specific sequence data items

My question is what do you think about extending response type of all workers to allow it to have additional data or do you have your own thoughts on this?

**Comment by vladislavkruglikov** [comment]
@WoosukKwon @Yard1 @youkaichao @cadedaniel what are your thoughts on this?

**Comment by cadedaniel** [comment]
What's the motivation to add this to the live-serving scenario? Unfortunately we are very latency constrained, so additional features must be justified, otherwise we will dismiss the latency reduction in live serving.

**Reply by vladislavkruglikov** [reply]
We want to use high efficiency of vllm in order to collect speculative decode states that would allow us to train better draft models for speculative decode. I propose to allow users to get internal states of speculative decode as an option, for example we might add special bool flag such as `return_hidden_states`

**Reply by vladislavkruglikov** [reply]
I found beautiful way to achieve what I want. I will attach pull request with that as soon as possible so that you can take a look

**Reply by vladislavkruglikov** [reply]
Created pull request #7549 still missing tests for new features but will add as soon as you would approve the general idea behind pull request, it seems looking good to me, does not break anything, just adds new things

**Reply by cadedaniel** [reply]
This seems to be a batch inference use-case. Why not provide a way to save these outputs to file in SpecDecodeWorker, so the serving path does not need any other overheads?

**Reply by vladislavkruglikov** [reply]
We need to have access to previous draft sequence and scores produced by target model in order to predict new draft in the subsequent iteration of speculative decode. I will create pull request for this feature after this pull request since this is blocker 

We serve vllm in cluster while access it from other cluster and thus we can not directly read filesystem of vllm in a nice way and generally I think that inference web server must be state less. We also need to receive speculative states after each iteration of decode so we can not just generate a huge file and save it and use it. It is kind of online use case and this is would be the fastest way to do so

**Reply by cadedaniel** [reply]
It's not a good idea to send these metrics to the user. There are two reasons:
* The collection of these metrics will never provide value for the average serving user, because they care about latency reduction and not internal acceptance rate metrics. Please correct me if I'm wrong.
* If the user is interested in improving their speculation strategy, then these metrics _do_ provide value. However the user can simply configure vLLM to emit these in the server logs. No need to pass the outputs through the server and back to the client.

I think it's fine to have an opt-in feature that logs these metrics you want in the SpecDecodeWorker. But it has to have ~zero impact on latency when it's disabled.

For perspective, python serialization operations can incur order 50µs, and GPU-sync operations (like those that exist in your PR) can occur 500µs+ of latency. These things add up and drown out speculative decoding's latency reduction.

**Reply by cadedaniel** [reply]
BTW, similar discussion happening here https://github.com/vllm-project/vllm/issues/7586

**Reply by vladislavkruglikov** [reply]
Metric is not the main purpose of this feature. Since our algorithm uses previous draft in order to predict next draft we need this feature. You can not implement this algorithm and do not store the previous drafts

What if I will add an additional flag such as `return_speculative_decode_states`  which will add `SpeculativeProposerSamplerOutput` and `SpeculativeScorerSamplerOutput` to the output and engine will check if that flag is set in order to start processing it. This will definitely help to remove GPU synchronization if you do not want to use additional features you do not want to pay for

**Reply by cadedaniel** [reply]
>Metric is not the main purpose of this feature. Since our algorithm uses previous draft in order to predict next draft we need this feature. You can not implement this algorithm and do not store the previous drafts

I am confused; why do you want to return these previous draft states to the user? If you want to improve your drafting live, e.g. https://arxiv.org/abs/2310.07177, you want to keep these in the SpecDecodeWorker.

**Reply by vladislavkruglikov** [reply]
If we would store speculative decode states in speculative decode worker this would definitely solve our problem. But one more thing that we need it so have ability to get internal states of speculative decoder as fast as possible. If we would store internal states as logs then it would be much slower to access it rather then directly receiving them as API response

**Reply by cadedaniel** [reply]
> If we would store speculative decode states in speculative decode worker this would definitely solve our problem

Great. Feel free to make a PR for this, or an RFC proposing the change. In general, if we can keep speculative decoding features inside the SpecDecodeWorker, it will be easier to merge.

> But one more thing that we need it so have ability to get internal states of speculative decoder as fast as possible. If we would store internal states as logs then it would be much slower to access it rather then directly receiving them as API response

Can you help me understand the use case better? Who is the user, why does the user need it, how fast do you need it, why can't it stay inside the SpecDecodeWorker.
